bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,423 @@
+        await self.symbol_get_event.wait()
+        log.debug("Refreshing documentation inventory...")
+        self.inventory_scheduler.cancel_all()
+
+        self.base_urls.clear()
+        self.doc_symbols.clear()
+        self.renamed_symbols.clear()
+        await self.item_fetcher.clear()
+
+        coros = [
+            self.update_or_reschedule_inventory(
+                package["package"], package["base_url"], package["inventory_url"]
+            ) for package in await self.bot.api_client.get('bot/documentation-links')
+        ]
+        await asyncio.gather(*coros)
+        log.debug("Finished inventory refresh.")
+        self.refresh_event.set()
+
+    async def get_symbol_embed(self, symbol_name: str) -> Optional[discord.Embed]:
+        """
+        Attempt to scrape and fetch the data for the given `symbol_name`, and build an embed from its contents.
+
+        If the symbol is known, an Embed with documentation about it is returned.
+
+        First check the DocRedisCache before querying the cog's `BatchParser`.
+        """
+        log.trace(f"Building embed for symbol `{symbol_name}`")
+        if not self.refresh_event.is_set():
+            log.debug("Waiting for inventories to be refreshed before processing item.")
+            await self.refresh_event.wait()
+
+        doc_item = self.doc_symbols.get(symbol_name)
+        if doc_item is None:
+            if symbol_name.count(" "):
+                # If an invalid symbol contains a space, check if the command was invoked
+                # in the format !d <symbol> <message>
+                symbol_name = symbol_name.split(" ", maxsplit=1)[0]
+                doc_item = self.doc_symbols.get(symbol_name)
+                if doc_item is None:
+                    log.debug("Symbol does not exist.")
+                    return None
+            else:
+                log.debug("Symbol does not exist.")
+                return None
+
+        self.bot.stats.incr(f"doc_fetches.{doc_item.package}")
+
+        with self.symbol_get_event:
+            markdown = await doc_cache.get(doc_item)
+
+        if markdown is None:
+            log.debug(f"Redis cache miss for symbol `{symbol_name}`.")
+            markdown = await self.item_fetcher.get_markdown(doc_item)
+            if markdown is not None:
+                await doc_cache.set(doc_item, markdown)
+            else:
+                markdown = "Unable to parse the requested symbol."
+
+        embed = discord.Embed(
+            title=discord.utils.escape_markdown(symbol_name),
+            url=f"{doc_item.url}#{doc_item.symbol_id}",
+            description=markdown
+        )
+        # Show all symbols with the same name that were renamed in the footer,
+        # with a max of 100 chars.
+        if symbol_name in self.renamed_symbols:
+            renamed_symbols = ', '.join(self.renamed_symbols[symbol_name])
+            footer_text = textwrap.shorten("Moved: " + renamed_symbols, 100, placeholder=' ...')
```

Comment:
I've moved the creation of the footer text before the get symbol await to avoid this.

Reply:
It's a fix because everything above is synchronous, right? Doesn't this seem quite fragile and implicit? If someone didn't know any better, they could introduce race conditions simply by moving re-arranging code, which is an ostensibly harmless thing (because none of the code has directly evident side effects). Just me rambling a bit. I don't have a good idea of how to deal with it nicely.